README.md
mempo-deb
=========

Mempo Linux - Deb - patches, utils, scripts for Debian packages

This will containt e.g. scripts to build fixes .deb packages.

See: https://wiki.debian.org/Mempo/mempo-deb

More docs, keys
=========

More doc (pgp keys, signatures, trust, donations address etc) are 
documented for now in other our repo, see:

https://github.com/mempo/deterministic-kernel/blob/master/doc/donate.txt
(quick copy: 
Dogecoin: DGGcXaNA6wUDn2GUnaPwmJz9WFAA9iLSWw
Noblecoin: 9iLGmzy365g98xXjUM81QJNdpVXKACXMyw
)

https://github.com/mempo/deterministic-kernel/blob/master/README.md
(quick copy:
You should verify data with both github.com by https/SSL or SSH downloads and you should check 
git tag -v tagname e.g.: git tag -v v0.1.26-rc3 to see if tag is signed by: pgp key 45953F23 rfree-mempo, 
pubkey here in doc/ full id: 21A5 9D31 7421 F02E C3C3 81F3 4623 E8F7 4595 3F23
)



changelog


mempo-deb (0.1.24) UNRELEASED; urgency=low

  * fix in xz no-lib version
  * deb-0.1.24.2
  * fix the patch format
  * deb-0.1.24.3
  * really fix the patch format (one of args could terminate arg list)
  * changed tags numbering
  * 0.1.24.5
  * docs
  * compression level for all compressors set always to -9 (gz, bz, xz)
  * 0.1.24.6
  * details (coding style)
  * zlib only compression, same compression strategy
  * added offline mode
  * dpkg strategy
  * require that zlib is installed
  * (other changes? check gitlog)
  * 0.1.24.9
  * require gettext version 0.18.3 (this is the current one, to make sure everyone uses same dpkg+gettext)
  * 0.1.24.10

 -- mempo <mempo@mempo.org>  Tue, 06 Jan 2014 11:30:00 +0200
	
mempo-deb (0.1.23+4) UNRELEASED; urgency=low

  * Fixed our dpkg (based on reproducible dpkg) to use sha256 for xz/liblzma
  * Switched version numbering to .5 instead +5.
  * deb-0.1.23.5
  * compression level always -9 (both lib lzma and command xz)
  * deb-0.1.23.7

 -- mempo <mempo@mempo.org>  Tue, 06 Jan 2014 11:30:00 +0200

mempo-deb (0.1.3) UNRELEASED; urgency=low

  * Change max key size in gpg-keygen
	* Fixed libpoco

 -- mempo <mempo@mempo.org>  Fri, 13 Dec 2013 13:00:00 +0200



pack/poco/poco-test.cpp
//You can test if you have pcre poco bug by compiling tis test program:
//
//$ g++ poco-test.cpp -o poco-test -l PocoFoundation
//
//and runnig it:
//
//$ echo 'Network.RTSP.Port="554"' | ./poco-test

#include <iostream>
#include <string>
#include <vector>

#include <Poco/RegularExpression.h>

int
main()
{
    std::string line;
    std::vector<std::string> subs;

    const Poco::RegularExpression re("([^=]+)=\"?([^\"]*)\"?");

    while (std::cin.good()) {
        std::getline(std::cin, line);
        if (!line.empty()) {
            std::cout << re.split(line, subs) << '\n';
        }
    }

    return 0;
}


pack/poco/go.sh
#!/bin/bash -e

# mempo-title: Fixed poco library against the lib pcre related bug
# mempo-prio: 2
# mempo-why: Program FMS was unusably broken
# mempo-bugfix-deb: http://bugs.debian.org/cgi-bin/bugreport.cgi?bug=671477

echo "Please run as ROOT (if needed): apt-get build-dep poco; apt-get install devscripts faketime"


#/tmp/tmpbuild
#/tmp/make1234
#12345678

dir_template="/tmp/makeXXXX"
dir="$(mktemp -d "$dir_template" )"
if [ -z "$dir" ] ; then echo "Problem creating temporary directory ($dir) from template ($dir_template)"; exit 1; fi
echo $dir


rm -rf "$dir"
mkdir -p "$dir" ; chmod 700 "$dir";
cp libpcre-8.13.patch checksums "$dir"
cd "$dir" # <---------

rm -rf build ; mkdir -p build ; cd build

apt-get source poco

cd poco-1.3.6p1
patch -p 0 < ../../libpcre-8.13.patch
faketime "2013-08-28 16:20:26" debuild -us -uc -B -j2

cd ..
FILES=*.deb
for f in $FILES
do
  echo "Extracting $f..."
  dpkg-deb -x $f out
done

echo "Checking sha512sum of builded libs"
sha512sum out/usr/lib/*.so > checksums-local
cp checksums-local /tmp/ # XXX

echo "TESTBASH:" # XXX
bash


echo "Differences:"
diff -Nuar checksums-to-verify ../checksums 

sha256deep -r "{$dir}/build/"

echo "Builded packages are in: {$dir}/build. After checksum verification install with dpkg -i *.deb"


pack/gnupg/build-patched.sh
#!/bin/bash

# <mempo>
# mempo-title: Change max key size in gpg-keygen
# mempo-prio: 2
# mempo-why: To improve pgp keys security
# mempo-bugfix-deb:
# </mempo>

# work in progress - XXX marks debug code

# TODO isolate this script into a common build-with-patch script
# TODO privacy: set commont timezone e.g. UTC and locale (just to be sure)

# http://mywiki.wooledge.org/BashFAQ/105/ ; http://mywiki.wooledge.org/BashFAQ/101

set -e

# warn: Print a message to stderr. Usage: warn "format" ["arguments"...]
warn() {
  local fmt="$1" ; shift ; printf "ERROR: $fmt\n" "$@" >&2
}
# Usage: some_command || die [status code] "message" ["arguments"...]
die() {
  local st="$?" ; if [[ "$1" != *[^0-9]* ]] ; then st="$1" ; shift ; fi
  warn "$@" ; exit "$st"
}

echo "Tools: checking prerequisites..."
DPKG_VER=$(dpkg-query -W --showformat='${Version}\n' dpkg)
DPKG_VER_NEEDED="1.17.5"

function show_dpkg_why {
        echo "We need dpkg version that packs files in same way, see http://tinyurl.com/pcrrvag and https://wiki.debian.org/ReproducibleBuildsKernel"
}

function show_mempo_contact {
        echo "~~ Problems, questions, suggestions or will to help us? ~~ Contact Mempo at IRC"
        echo "IRC channel #mempo on irc.oftc.net (tor allowed), irc2p (i2p2.de then localhost 6668) or irc.freenode.org."
        echo "We will gladly help fellow Hackers and security researchers."
}

echo " * Dpkg version is $DPKG_VER (version >= $DPKG_VER_NEEDED is recommended)"

if [[ 2 > 3 ]] ; 
then
	. dpkg-vercomp.sh
	vercomp $DPKG_VER $DPKG_VER_NEEDED
	case $? in
					2)
					echo "Wrong DPKG version..." ;
					echo "If you want to force and try despite this problem, edit this script that shows this error."
					show_dpkg_why
					show_mempo_contact
					echo
					echo "On Debian (wheezy) the SOLUTION is to install dpkg in version from jessy (download sources, build only this one package, install it), search for more info on our Wiki."
					echo
					exit 1
					;;
	esac
else
	echo "Skipping check of dpkg version."
fi

base_dir="$(pwd)" ; [ -z "$base_dir" ] && die "Could not get pwd ($base_dir)" # basic pwd (where our files are)
# echo "Our base_dir is $base_dir [PRIVACY]"  # do not print this because it shows user data

echo ; echo "Please run as ROOT (if needed): apt-get build-dep gnupg; apt-get install devscripts faketime" ; echo

#/tmp/tmpbuild
#dir_template="/tmp/build-XXXXXX"
#build_dir="$(mktemp -d "$dir_template" )" # dir to build in
build_dir=/tmp/build-ZZZZZZ-gnupg
if [ -z "$build_dir" ] ; then die "Problem creating temporary directory ($build_dir) from template ($dir_template)"; fi
echo "Building in $build_dir"

rm -rf "$build_dir" || die "Can't delete build_dir ($build_dir)"
mkdir -p "$build_dir" || die "Creating build_dir ($build_dir)"
chmod 700 "$build_dir" || die "While chmod build_dir ($build_dir)" # create build dir

# copy files to build dir
cp genlongkey.patch "$build_dir"
cp checksums_expected "$build_dir"

# ===================================================================
# --- operations inside build dir
function execute_in_build_dir() {
cd "$build_dir" || die "Can not enter the build_dir ($build_dir)" # <---------

rm -rf build ; mkdir -p build ; cd build # recreate build directory

if [[ $1 == "offline" ]]
then
  echo "Building in OFFLINE mode, using provided sources"
  cp $base_dir/src/* . # XXX
  sha512sum *.gz
  sha512sum *.dsc
  echo "Does above checksums of the SOURCE file look correct? Press Ctrl-C to cancel or ENTER to conitnue"
	echo "(Verify this checksums from external trusted source, although this files ARE EMBBED in the code download, e.g. should be embbed in git clone / git exported bundle here in OFFLINE mode, so should be fine if you trust this script)"
  read _
  dpkg-source -x *.dsc
else
  echo "Downloading sources using apt-get source (from network)"
	set -x
  apt-get source gnupg || die "Can not download the sources" # <--- download
	set +x
	echo "This download should be secured thanks to apt-get security"
fi


#cd patch -p0 < genlongkey.patch
patch -p 0 < "$base_dir/genlongkey.patch"
cd gnupg-1.4.12
tmp1="$(mktemp "/tmp/build-data-XXXXXX")" || die "temp file"
[ ! -w "$tmp1" ] && die "use temp ($tmp1)"
cat "$base_dir/changelog" debian/changelog > "$tmp1" || die "Writting debian/rules"
mv "$tmp1" debian/changelog || die "Moving updated debian/rules"

DEB_BUILD_TIMESTAMP="2013-08-28 16:20:26"

faketime "2013-08-28 16:20:26" dpkg-buildpackage -us -uc -B || die "Failed to build"

cd ..
sha512sum *.deb > checksums_local

echo "Differences:"
DIFFER=$(diff -Nuar ../checksums_expected checksums_local)
if [ -z "$DIFFER" ]; then
    echo -e "\e[42mNO DIFFERENCES, ALL OK\e[0m"
else
    echo -e "\e[41mWARNING! CHECKSUMS ARE DIFFERENT\e[0m"
    echo "$DIFFER"
fi
echo "Builded packages are in: $build_dir/build. After checksum verification install with dpkg -i *.deb"

}
# inside build dir
# ===================================================================

execute_in_build_dir $1


pack/checksumdb/src/util/check_debian_iso.sh
#!/bin/sh
#
# check_debian_iso, copyright 2011 Thomas Schmitt <scdbackup@gmx.net>
# License:   GPLv2 or later
# Tested on: Little-endian GNU/Linux with bash
#            Little-endian FreeBSD-8 with sh and "md5 -q"
#            Little-endian Solaris 5.11 with ksh93
#            Big-endian    GNU/Linux with bash
#
# Modified by rfree@mempo.org to be general ISO 9660 checksum 
# See github https://github.com/mempo/mempo-deb/tree/master/pack/checksumdb and search
# for check_debian_iso.sh for updated version 

prog=`basename "$0"`

usage() {
  echo "Usage: $prog Checksum_file [U:]Item [Image_file] [Checksum_command]" >&2
  echo "" >&2
  echo "This program can calculate checksum of DATA (skipping the empty parts)" >&2
  echo "of a CD, DVD or other ISO 9660 image: *.iso file, /dev/dvd device etc" >&2
  echo "" >&2
  echo "Or use with debian checksum files:" >&2
  echo "Reads the checksum of a Debian installation image from Checksum_file" >&2
  echo "and compares it with the ISO 9660 image in Image_file. Suitable" >&2
  echo "for verifying optical media, because trailing garbage is ignored." >&2
  echo "The Item in the Checksum_file is depicted either by its complete" >&2
  echo "file name (e.g. debian-6.0.3-amd64-CD-1.iso) or by a text piece" >&2
  echo "between '-' and '.iso' in the file name. The first match is used." >&2
  echo "Text pieces for debian-update images must be prefixed by 'U:'." >&2
  echo "If no Image_file is given, then the item file name is used instead." >&2
  echo "Checksum_command is normally deduced from Checksum_file name." >&2
  echo "It must read data from standard input and its first word written" >&2
  echo "to standard output must be the checksum. Default commands are" >&2
  echo "md5sum, sha1sum, sha256sum, sha512sum." >&2
  echo "" >&2
  echo "Examples:" >&2
  echo "" >&2
  echo "  $prog  x  x  /dev/dvd sha512sum" >&2
  echo "" >&2
  echo "  $prog  MD5SUMS  debian-6.0.3-amd64-netinst.iso" >&2
  echo "  $prog  MD5SUMS  netinst" >&2
  echo "  $prog  MD5SUMS  debian-6.0.3-amd64-DVD-1.iso  /dev/dvd" >&2
  echo "  $prog  MD5SUMS  1  /dev/dvd" >&2
  echo "  $prog  MD5SUMS  2  /dev/dvd" >&2
  echo "  $prog  MD5SUMS  U:1  /dev/dvd" >&2
  echo "  $prog  MD5SUMS  kde-CD-1  /dev/cdrom" >&2
  echo "  $prog  SHA512SUMS  businesscard  /dev/cdrom" >&2
  echo "  $prog  MD5SUMS  1 /dev/cd0 'md5 -q'" >&2
}

if test -z "$1" -o "$1" = "-h" -o "$1" = "--help" -o -z "$2"
then
  usage
  exit 1
fi
sums="$1"
vol="$2"
file="$3"

checksummer=md5sum
if test -n "$4"
then
  checksummer="$4"
else
  base=`basename "$sums"`
  echo "Guessing checksummer algorithm to use: $base"
  if test "$base" = "SHA1SUMS"
  then
    checksummer=sha1sum
  elif test "$base" = "SHA256SUMS"
  then
    checksummer=sha256sum
  elif test "$base" = "SHA512SUMS"
  then
    checksummer=sha512sum
  else 
    echo "Can not get the name of checksum... pass it as 4th argument. Run program with no options to see usage help."
    exit 1
  fi
fi

update=""
update_v="-v"
use_fgrep=""
if echo "$vol" | grep '^debian-.*\.iso$' >/dev/null
then
  use_fgrep=1
elif echo "$vol" | grep '^U:' >/dev/null
then
  update=" update"
  update_v=""
  vol=`echo "$vol" | sed -e 's/^U://'`
fi

if test -n "$use_fgrep"
then
  line_from_list=`fgrep "$vol" "$sums" | head -1`
else
  line_from_list=`grep '.*-'"$vol"'\.iso$' "$sums" | grep $update_v "update" | head -1`
fi
sum_from_list=`echo "$line_from_list" | awk '{print $1}'`
name_from_list=`echo "$line_from_list" | awk '{print $2}'`

if test -z "$sum_from_list"
then
  if test -n "$use_fgrep"
  then
    echo "Could not find item '${vol}' in '$sums'" >&2
  else
    echo "Could not find$update item '.*-${vol}.iso' in '$sums'" >&2
  fi
  #exit 4
fi

if test -z "$file"
then
  file="$name_from_list"
fi

# Logical block size is assumed with 2048 bytes. Neither genisoimage
# nor xorriso produce other sizes, and even the Linux kernel seems to
# have this size hardcoded.

# At byte 16 * 2048 starts the Primary Volume Descriptor (superblock)
# of the image. The magic number values should be ECMA-119 Volume
# Descriptor Type 0x01 and Standard Identifier "CD001".
# The way how these 6 bytes group to 16-bit words indicates the byte
# sex (endianness) of the local machine. The ECMA-119 Volume Space Size
# is stored as little-endian 32-bit number at PVD byte 80, and as big-endian
# 32-bit number at PVD byte 84.

# od -d is used because it guarantees unsigned integer of predictable
# size. Formats -i and -l depend on sizeof(int).

echo "Reading magic:"
dd if="$file" bs=2048 skip=16 count=1 | hexdump -C # test it

magic=`(dd if="$file" bs=2048 skip=16 count=1 |
        dd bs=1 count=6 | od -x | head -1 | \
        awk '{print $2 " " $3 " " $4}') 2>/dev/null`
if test "$magic" = "4301 3044 3130"
then
  lo=`(dd if="$file" bs=2048 skip=16 count=1 | \
       dd bs=1 skip=80 count=2 | od -d | head -1 | \
       awk '{print $2}') 2>/dev/null`
  hi=`(dd if="$file" bs=2048 skip=16 count=1 | \
       dd bs=1 skip=82 count=2 | od -d | head -1 | \
       awk '{print $2}') 2>/dev/null`
elif test "$magic" = "0143 4430 3031"
then
  lo=`(dd if="$file" bs=2048 skip=16 count=1 | \
       dd bs=1 skip=86 count=2 | od -d | head -1 | \
       awk '{print $2}') 2>/dev/null`
  hi=`(dd if="$file" bs=2048 skip=16 count=1 | \
       dd bs=1 skip=84 count=2 | od -d | head -1 | \
       awk '{print $2}') 2>/dev/null`
elif test -e "$file"
then
  echo "Does not look like an ISO 9660 filesystem: '$file' magic='$magic'" >&2
  exit 2
else
  echo "File not found: '$file'" >&2
  exit 5
fi

# The two 16 bit numbers, which are of the appropriate byte sex,
# get combined to a 32 bit number.
blocks=`expr $lo + $hi '*' 65536`
echo "Calculated media size to blocks=$blocks (lo=$lo hi=$hi)"

echo "Piping $blocks blocks of '$file' through '$checksummer'" >&2
echo "to verify checksum list item '$name_from_list'." >&2

sum_from_file=`dd if=$file bs=2048 count=$blocks | $checksummer | head -1 | awk '{print $1}'`

if test "$sum_from_list" = "$sum_from_file"
then
  echo "Ok: '$file' matches$update '$name_from_list' in '$sums'"
else
  echo "Found:     $sum_from_file" >&2
  echo "Expected:  $sum_from_list" >&2
  echo "MISMATCH: '$file' checksum differs from '$name_from_list' in '$sums'"
  exit 3
fi
exit 0



pack/dpkg/build-and-install-locally.sh
#!/bin/bash

# <mempo>
# mempo-title: Add posiibility to set timestamp in ar deb archive
# mempo-prio: 2
# mempo-why: To make deterministic deb package possible
# mempo-bugfix-deb:
# </mempo>

# work in progress - XXX marks debug code

# TODO isolate this script into a common build-with-patch script
# TODO privacy: set commont timezone e.g. UTC and locale (just to be sure)

# http://mywiki.wooledge.org/BashFAQ/105/ ; http://mywiki.wooledge.org/BashFAQ/101

set -e

# warn: Print a message to stderr. Usage: warn "format" ["arguments"...]
warn() {
  local fmt="$1" ; shift ; printf "ERROR: $fmt\n" "$@" >&2
}
# Usage: some_command || die [status code] "message" ["arguments"...]
die() {
  local st="$?" ; if [[ "$1" != *[^0-9]* ]] ; then st="$1" ; shift ; fi
  warn "$@" ; exit "$st"
}

#~base_dir="$(pwd)" ; [ -z "$base_dir" ] && die "Could not get pwd ($base_dir)" # basic pwd (where our files are)
# echo "Our base_dir is $base_dir [PRIVACY]"  # do not print this because it shows user data

echo "This script will download, build and install locally ($HOME/.local) dpkg with Lunar's deterministic patches (http://bugs.debian.org/cgi-bin/bugreport.cgi?bug=719845#54)" ; echo
echo "Please run as ROOT (if needed): apt-get install git build-dep gnupg; apt-get install devscripts autoconf automake flex" ; echo
echo ""

gettext_ver=$( LC_ALL=C dpkg -s gettext | grep 'Version' | head -n 1 | sed -e "s/Version: \([^ ]*\).*/\1/" | cut -d'-' -f1 )
echo " * gettext version=$gettext_ver"

. dpkg-vercomp.sh 

ver_what='gettext'; ver_have=$gettext_ver ; ver_need="0.18.3"
set +e ; vercomp $ver_have $ver_need ; err=$? ; set -e 
case $err in
  2) echo ; echo "ERROR: wrong version of $ver_what"
                echo "We have $ver_have while we need $ver_need" ; 
								echo "Because we need GNU gettext >= 0.18.2, please add \"http://YOURMIRROR.debian.org/debian wheezy-backports\" to /etc/apt/sources.list and run:"
								echo "aptitude update"
								echo "aptitude install -t wheezy-backports gettext autopoint"
								echo "See http://wiki.debian.org/Mempo/ for help"
                exit 1;
        ;;
esac
echo "ok"
echo " * Using $ver_what version $ver_have >= $ver_need - OK"

rm -rf dpkg

dpkg_pu_git_version="9673d63303211fdefe650f2974d35d326929d0fd" # the version to download

if [[ $1 == "offline" ]]
then
  echo "Building in OFFLINE mode, using provided sources"
  mkdir dpkg
  tar -xzvf src/1.17.7~reproducible1.tar.gz -C dpkg
  echo ""
  cd dpkg
else
  git clone https://alioth.debian.org/anonscm/git/reproducible/dpkg.git || die "Can't clone dpkg repository"
  cd dpkg
  git checkout pu/reproducible_builds || die "Can not checkout the branch we need"
  git checkout $dpkg_pu_git_version || die "Can not checkout this git version that we need - $dpkg_pu_git_version"
  echo "Checking repository reference number"
  gitver="$(git show-ref --hash --heads)" || die "Can't take repository reference number!"

  if [[ "$gitver" == "$dpkg_pu_git_version" ]] ; then
    echo "OK GIT VERSION: $gitver"
    else
    die "Github repository reference doesn't match! We got gitver=$gitver instead of expected dpkg_pu_git_version=$dpkg_pu_git_version"
  fi
fi

patch -p 0 < ../set-version-manually-because-no-tags.patch
patch -p 1 < ../force-all-compressors-level-9-and-sha256-for-xz.patch
patch -p 1 < ../check-if-zlib-is-installed.patch
patch -p 1 < ../set-strategy-manually.patch

autoreconf -f -i
./configure --prefix=$HOME/.local
make
make install

echo "======================================="
echo "dpkg build and installed in $HOME/.local"
